tests/decap/conftest.py

```diff
@@ -0,0 +1,27 @@
+
+# Decap pytest arguments
+def pytest_addoption(parser):
+
+    parser.addoption(
```

Comment:
args are going to grow per feature and overall. I expect it would be hard to know what each arg/param does. What about grouping them into subgroups?

https://github.com/Azure/sonic-mgmt/pull/1713/files#diff-214049cbadc3581f96630166561f2436R13

Reply:
Good suggestion! Thanks, I'll update.

Reply:
It is a copy and paste error. Here is the link:

https://github.com/Azure/sonic-mgmt/pull/1713/files#diff-36e2e7cb17577532cab5be133af2e37fR587

ptf runner would be hard to maintain if we want to include all possible parameters. Also, it does nothing important perse and ptf command could be invoked directly.  Eventually ptfrunner would be obsolete once Ansible is.

Reply:
I think it would be easy to enhance the ptf_runner to accept any parameters. It can be considered as a helper function for building ptf cli string with many parameters and run it. Without this, it would be more tedious to run ptf scripts.

Reply:
can you please elaborate more on the making ptf cli easier? The point is that ptf has many argument as could be seen [here](https://github.com/p4lang/ptf/blob/master/ptf#L51).  Eventually, we will need to add all those arg to the ptf runner with or withoutdefault. Whereas if we invoke ptf directly via ansible module, each test case would include only those parameters of interest in the argv array. I do not see the overhead to be much and ptf runner is providing little functionality.

Reply:
I personally prefer having the helper as Xin is suggesting. I'd rather we define a clear interface in Python for setting up ptf tests than expecting every client to dig into the code to figure out what exact string they need, how to setup the format parameters, etc.

Reply:
I am not sure why digging is bad to start with. In either case, one would need to look at ptfrunner to see what new (map) string might be for the corresponding ptf parameter (local digging) The other case is when writing new test/feature, one need to dig new parameters any way in order to learn how they work and use them. And then Ansible does give a way to compose the cli command when using argv.

Reply:
@tahmed-dev To run a PTF script on the PTF docker, we need to run a command line like below from PTF shell:
```
$ ptf --test-dir ptftests IP_decap_test.DecapPacketTest --platform-dir ptftests --qlen=1000 --platform remote -t 'fib_info='"'"'/root/fib_info.txt'"'"';testbed_type='"'"'t1'"'"';ttl_mode='"'"'
pipe'"'"';vlan_ipv6='"'"''"'"';dscp_mode='"'"'pipe'"'"';lo_ipv6='"'"'None'"'"';router_mac=u'"'"'52:54:00:12:34:56'"'"';lo_ip='"'"'10.1.0.1'"'"';outer_ipv6=False;outer_ipv4=True;inner_ipv4=True;inner_ipv6=False
;vlan_ip='"'"''"'"';src_ports='"'"'0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31'"'"'' --relax --debug info --log-file /tmp/decap.2020-06-01-06:21:30.log
```
The ptf_runner can build this command line string from the arguments passed to it:
```
    ptf_runner(ptfhost,
               "ptftests",
               "IP_decap_test.DecapPacketTest",
                platform_dir="ptftests",
                params={"testbed_type": testbed['topo']['type'],
                        "outer_ipv4": setup_info["outer_ipv4"],
                        "outer_ipv6": setup_info["outer_ipv6"],
                        "inner_ipv4": setup_info["inner_ipv4"],
                        "inner_ipv6": setup_info["inner_ipv6"],
                        "lo_ip": setup_info["lo_ip"],
                        "lo_ipv6": setup_info["lo_ipv6"],
                        "vlan_ip": setup_info["vlan_ip"],
                        "vlan_ipv6": setup_info["vlan_ipv6"],
                        "dscp_mode": setup_info["dscp_mode"],
                        "ttl_mode": setup_info["ttl_mode"],
                        "src_ports": setup_info["src_ports"],
                        "router_mac": setup_info["router_mac"],
                        "fib_info": FIB_INFO_DEST,
                        },
                qlen=PTFRUNNER_QLEN,
                log_file=log_file)
```

I just need to prepare these parameters and pass them to the ptf_runner function. Then the ptf_running function will concatenate them into a command line string, for example `cmd` and run it using `ptfhost.shell(cmd)`. Then return the result. If I do not use ptf_runner, I have to concatenate the parameters into a command line string in the test scripts by myself.

>Whereas if we invoke ptf directly via ansible module, each test case would include only those parameters of interest in the argv array.

Do you mean there is a dedicated ansible module which can run the PTF test? Then we just need to call the module and pass the parameters to it? I searched ansible docs and couldn't find such an ansible module.

Reply:
After checking tests/common/devices.py:PTFHost, it is defined as below:
```
class PTFHost(AnsibleHostBase):
    """
    @summary: Class for PTF

    Instance of this class can run ansible modules on the PTF host.
    """
    def __init__(self, ansible_adhoc, hostname):
        AnsibleHostBase.__init__(self, ansible_adhoc, hostname)

    # TODO: Add a method for running PTF script
```

There is TODO comment there "# TODO: Add a method for running PTF script". Now it's clear, we can implement a method for the PTFHost class for running PTF scripts. The method basically do the same like the ptf_runner. Then the ptf_runner helper function can be deprecated. Is this what your mean @tahmed-dev ?

Reply:
Thanks @wangxin. It is good to have a method in the PTFHost to invoke cmd/shell. I'd recommend shying away from bloating it with the same contents as ptf_runner. We could use the argv format of the shell and invoke the command as:

```
ptfhost.shell(
            argv = [
                "ptf",
                "--test-dir",
                "saitests",
                testCase,
                "--platform-dir",
                "ptftests",
                "--platform",
                "remote",
                "-t",
                ";".join(["{}={}".format(k, repr(v)) for k, v in testParams.items()]),
                "--disable-ipv6",
                "--disable-vxlan",
                "--disable-geneve",
                "--disable-erspan",
                "--disable-mpls",
                "--disable-nvgre",
                "--log-file",
                "/tmp/{0}.log".format(testCase),
            ],
            chdir = "/root",
        )
```

Args that are fixed could then be moved to the PTFHost and the class could append them to the argv. Also, unpacking of test_params could also be a util/method in PTFHost method. I do not see a need to to stick to cmd form of the shell command. Ansible shell command does the packing and unpacking for argv.

Reply:
@tahmed-dev Thanks for your great suggestions! I'll create another PR to improve this.